### PR TITLE
[GPU] Reinterpret from 1 dim mem to 0 dim mem instead of allocating 0 bytes layout to OpenCL

### PR DESCRIPTION
**Pull request by kelvinchoi-intel**
### Details:
 - *Requesting 0 bytes memory allocation to OpenCL returns error -61. Plugin should not request 0 bytes memory allocation.*

### Tickets:
 - *143586*
 - *143335*
 - *142909*

### CODE REVIEW

**Comment by ahnyoung-paul** [comment]
@sshlyapn Could you please review this PR to allocate 1 byte for zero byte allocation? thanks.

File: src/plugins/intel_gpu/src/runtime/ocl/ocl_engine.cpp [hunk 1]
```diff
@@ -174,23 +174,41 @@ memory::ptr ocl_engine::allocate_memory(const layout& layout, allocation_type ty
 
     check_allocatable(layout, type);
 
+    auto zero_bytes_layout = false;
+    auto non_zero_layout = layout;
+    if (layout.bytes_count() == 0) {
+        cldnn::layout zero_dim_layout = layout;
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/runtime/ocl/ocl_engine.cpp:180]: Why do you create layout copy?

File: src/plugins/intel_gpu/src/runtime/ocl/ocl_engine.cpp [hunk 2]
```diff
@@ -174,23 +174,41 @@ memory::ptr ocl_engine::allocate_memory(const layout& layout, allocation_type ty
...
+        auto mem_ps = zero_dim_layout.get_partial_shape();
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/runtime/ocl/ocl_engine.cpp:181]: No need to query real partial shape, just use `{1}`

File: src/plugins/intel_gpu/src/runtime/ocl/ocl_engine.cpp [hunk 3]
```diff
@@ -174,23 +174,41 @@ memory::ptr ocl_engine::allocate_memory(const layout& layout, allocation_type ty
 
     check_allocatable(layout, type);
 
+    auto zero_bytes_layout = false;
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/src/runtime/ocl/ocl_engine.cpp:177]: [random spot] I think previously we tried to avoid attempts to allocate 0 bytes on external code as it's something not expected (you can't do anything with buffer after that). Is it some special case? Maybe we can do a fix on caller code side?

File: src/plugins/intel_gpu/src/runtime/ocl/ocl_engine.cpp (hunk 1 above)
Comment (kelvinchoi-intel) [reply src/plugins/intel_gpu/src/runtime/ocl/ocl_engine.cpp]: updated

File: src/plugins/intel_gpu/src/runtime/ocl/ocl_engine.cpp (hunk 2 above)
Comment (kelvinchoi-intel) [reply src/plugins/intel_gpu/src/runtime/ocl/ocl_engine.cpp]: updated

File: src/plugins/intel_gpu/src/runtime/ocl/ocl_engine.cpp (hunk 3 above)
Comment (kelvinchoi-intel) [reply src/plugins/intel_gpu/src/runtime/ocl/ocl_engine.cpp]: I've updated to move the alloc-1dim-n-reinterpret-to-0dim code to util code instead of modify allocate_memory() in ocl_engine directly.
Loop input node like parameter type can have {0, 0} shape for input/output and the alloc-1dim-n-reinterpret-to-0dim code is necessary for that. Or any better idea?

**(snippet from TFLite model1_ov.xml)**
<layer id="7" name="model/bidirectional/forward_lstm_1/while/Placeholder_1" type="Parameter" version="opset1">
    <data shape="0,0" element_type="f32" />
    <rt_info>
        <attribute name="fused_names" version="0" value="model/bidirectional/forward_lstm_1/while/Placeholder_1" />
        <attribute name="old_api_map_element_type" version="0" value="f16" />
    </rt_info>
    <output>
        <port id="0" precision="FP32">
            <dim>0</dim>
            <dim>0</dim>
        </port>
    </output>
</layer>

File: src/plugins/intel_gpu/include/intel_gpu/plugin/common_utils.hpp [hunk 4]
```diff
@@ -111,6 +112,21 @@ inline ov::Shape predict_shape(const std::string& name, const cldnn::layout layo
     return layout.get_shape();
 }
 
+inline cldnn::memory::ptr allocate_memory_evenif_zero_bytes(cldnn::engine& _engine, const cldnn::layout& layout, cldnn::allocation_type type, bool reset = true) {
+    if (layout.bytes_count() == 0) {
+        auto non_zero_layout = cldnn::layout({1}, layout.data_type, layout.format);
+        auto res = _engine.allocate_memory(non_zero_layout, type, reset);
```
Comment (sshlyapn) [inline src/plugins/intel_gpu/include/intel_gpu/plugin/common_utils.hpp:118]: I think we can ignore reset value here in case of "empty buffer"

Comment (kelvinchoi-intel) [reply src/plugins/intel_gpu/include/intel_gpu/plugin/common_utils.hpp]: Updated.

File: src/plugins/intel_gpu/include/intel_gpu/plugin/common_utils.hpp [hunk 5]
```diff
@@ -111,6 +112,26 @@ inline ov::Shape predict_shape(const std::string& name, const cldnn::layout layo
     return layout.get_shape();
 }
 
+inline cldnn::memory::ptr allocate_memory_evenif_zero_bytes(cldnn::engine& _engine,
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/include/intel_gpu/plugin/common_utils.hpp:115]: Why did you move this logic out of engine::allocate_memory() ? My previous point was about not calling allocate_memory for 0 bytes at all if possible (i.e. if the buffer is not used). But in case if you actually need to save a scalar in memory, then I think we can have such logic within common allocate_memory()

Comment (kelvinchoi-intel) [reply src/plugins/intel_gpu/include/intel_gpu/plugin/common_utils.hpp:115]: Zero bytes memory seems to be not common case but Loop op requires it where handles input/output/backedge memory.
I agreed original allocate_memory() function should not modify for some specific case, so I decided it need to be handled as util function to reinterpret. The util function could be moved to ocl_engine with current function name. Better idea?
